File: python/fizzbuzz/fizzbuzz_test.py
# pylint: disable=unused-variable
"""Unit tests for FizzBuzz."""

# import the code to be tested
from fizzbuzz import fizz

# import utilities for writing tests
from pytest import raises

def describe_a_fizzbuzz_program_that():
  """A program to play the FizzBuzz game."""

  def has_a_smoke_test():
    """Make sure our testing infrastructure is working."""
    assert True == True

  def describes_a_fizz_function_that():

    def throws_an_error_if_fizz_has_no_input():
      """Makes sure fizz() throws an error if there is no input"""
      with raises(Exception) as err_info:
        fizz() # pylint: disable=no-value-for-parameter
      assert err_info.type == TypeError
      assert 'missing 1 required positional argument' in str(err_info.value)

    def outputs_fizz_if_an_input_is_numeric_and_a_multiple_of_3():
      """
        Checks to see if the input `x` is numeric and a multiple of 3.
        If it is, it outputs 'fizz'.
        Otherwise, it outputs the input.
      """
      assert fizz(3) == 'fizz'      # multiple of 3
      assert fizz(4) == 4           # non-multiple of 3
      assert fizz(0) == 'fizz'      # zero
      assert fizz(15) == 'fizz'     # multiple of 3 and 5
      assert fizz(-3) == 'fizz'     # negative multiple of 3
      assert fizz(-2) == -2         # negative non-multiple of 3
      assert fizz(3.5) == 3.5       # non-integer
      assert fizz('buzz') == 'buzz' # non-numeric

File: js/fizzbuzz/__tests__/fizzbuzz_test.js
// import the code to be tested
const { fizz, buzz, fibu, play } = require('../fizzbuzz') // the "old" safe way to import stuff

describe('A FizzBuzz program', () => {
  it('has a smoke test', () => {
    expect(true).toBe(true)
    expect(true).not.toBe(false)
  })

  describe('A fizz() function', () => {
    it('throws an error if there is no input', () => {
      expect(() => { fizz() }).toThrow('there is no input')
      expect(() => { fizz(3) }).not.toThrow()
    })

    it('returns "Fizz" if x is numeric and a multiple of 3', () => {
      expect(fizz(3)).toBe('Fizz')
      expect(fizz(4)).toBe(4)
      expect(fizz(0)).toBe('Fizz')
      expect(fizz(15)).toBe('Fizz')
      expect(fizz(-3)).toBe('Fizz')
      expect(fizz(-2)).toBe(-2)
      expect(fizz(3.5)).toBe(3.5)
      expect(fizz('Buzz')).toBe('Buzz')
    })
  })

  describe('A buzz() function', () => {
    it('throws an error if there is no input', () => {
      expect(() => { buzz() }).toThrow('there is no input')
      expect(() => { buzz(5) }).not.toThrow()
    })

    it('returns "Buzz" if x is numeric and a multiple of 5', () => {
      expect(buzz(5)).toBe('Buzz')
      expect(buzz(4)).toBe(4)
      expect(buzz(0)).toBe('Buzz')
      expect(buzz(15)).toBe('Buzz')
      expect(buzz(-5)).toBe('Buzz')
      expect(buzz(-2)).toBe(-2)
      expect(buzz(5.5)).toBe(5.5)
      expect(buzz('Fizz')).toBe('Fizz')
    })
  })

  describe('A fibu() function', () => {
    it('throws an error if there is no input', () => {
      expect(() => { fibu() }).toThrow('there is no input')
      expect(() => { fibu(15) }).not.toThrow()
    })

    it('returns "FizzBuzz" if x is numeric and a multiple of 15', () => {
      expect(fibu(15)).toBe('FizzBuzz')
      expect(fibu(4)).toBe(4)
      expect(fibu(0)).toBe('FizzBuzz')
      expect(fibu(-15)).toBe('FizzBuzz')
      expect(fibu(-2)).toBe(-2)
      expect(fibu(15.5)).toBe(15.5)
      expect(fibu('Fizz')).toBe('Fizz')
    })
  })

  it('can play FizzBuzz', () => {
    expect(play(1, 15)).toStrictEqual([
      1, 2, 'Fizz', 4, 'Buzz', 'Fizz', 7, 8, 'Fizz', 'Buzz', 11, 'Fizz', 13, 14, 'FizzBuzz'
    ])
  })
})